MELOSYS-3535 - Vedlegg opprett BUC

sigbjorn-myhre opened the pull request:
Legger til vedlegg-velger ved opprettelse av BUC. Har skrevet om ducks for dokument til typescript, samt deler av api-kallene.

sigbjorn-myhre committed:
Henter dokumentoversikt til redux og skriver om ducks til ts

sigbjorn-myhre committed:
Legger til valg av vedlegg ved opprettelse av BUC

JohanSakshaug commented on src/@types/dokument.ts:
```diff
@@ -0,0 +1,29 @@
+
+export interface FysiskDokument extends Dokument {
+  id: string,
+  dokumentID: string,
```
Arves ikke dokumentID, tittel og logiskeVedlegg fra `Dokument` her?

sigbjorn-myhre replied:
Jo, litt copy paste-feil her 😅

JohanSakshaug commented on src/ducks/dokumenter/operations.ts:
```diff
@@ -7,34 +7,26 @@
-    FEILET: Types.FEILET,
-    PENDING: Types.PENDING,
-  });
+export async function opprettDokument(behandlingID: number, dokumenttypeKode: string, dokument: BrevbestillingDto): Promise<void> {
```
Trenger ikke async uten await her

JohanSakshaug replied:
Er denne funksjonen egentlig nødvendig? Kunne man ikke brukt `Api.Dokumenter.dokument.opprett` direkte?

sigbjorn-myhre replied:
Ja, kan bare bruke den direkte.

JohanSakshaug commented on src/ducks/dokumenter/operations.ts:
```diff
@@ -7,34 +7,26 @@
 }
 
-async function getObjectURL(response) {
+async function getObjectURL(response: any): Promise<string> {
```
Skal gå fint å bruke `Response` i stedet for any her(typen er innebygd i ts).

JohanSakshaug commented on src/ducks/dokumenter/operations.ts:
```diff
@@ -7,34 +7,26 @@
 }
 
-export async function forhandsvisBrev(behandlingID, dokumenttypeKode, data) {
+export async function forhandsvisBrev(behandlingID: number, dokumenttypeKode: any, data: {} | any) {
```
dokumenttypeKode er vel string her?

data kunne vel hatt en type, som kunne vært gjenbrukt i pdfLenkeListe. Navneforslag: BrevData, under domenemodul, for eksempel?

JohanSakshaug commented on src/ducks/dokumenter/operations.ts:
```diff
@@ -49,7 +41,7 @@ export async function forhandsvisBrev(behandlingID, dokumenttypeKode, data) {
 }
 
-export async function forhandsvisSed(behandlingID, sedType, data) {
+export async function forhandsvisSed(behandlingID: number, sedType: string, data: any) {
```
Samme som kommentar over for data. SedData?

sigbjorn-myhre replied:
Ja, men tror jeg vil gå for `SedPdfData`. Det er navnet vi har i backend, og tror vi kommer til å få noen navnekonflikter senere med `SedData`.

JohanSakshaug replied:
👍  BrevPdfData også da?

JohanSakshaug commented on src/ducks/dokumenter/operations.ts:
```diff
@@ -67,6 +59,18 @@ export async function forhandsvisSed(behandlingID, sedType, data) {
 
-export function resetDokument() {
-  return dispatch => (dispatch(Actions.resetDokment()));
+export function resetDokument(): (dispatch: ThunkDispatch<RootState, unknown, Types.Action>) => Types.Action {
```
Kan bruke `AppThunk<Types.Action, Types.Action>` som returtype på `resetDokument.`

JohanSakshaug commented on src/ducks/dokumenter/operations.ts:
```diff
@@ -67,6 +59,18 @@ export async function forhandsvisSed(behandlingID, sedType, data) {
+  return (dispatch: ThunkDispatch<RootState, unknown, Types.Action>) => (dispatch(Actions.reset()));
+}
+
+export function hentDokumentOversikt(saksnummer: string): AppThunk<Promise<Types.Action>, Types.Action> {
```
👍

JohanSakshaug commented on src/ducks/dokumenter/selectors.ts:
```diff
@@ -0,0 +1,87 @@
+import MKV from '../../melosyskodeverk';
+
+export const dokumenterSelector = createSelector(
+  (state: { dokumenter: { data: { dokumentOversikt: DokumentOversikt[] } } }) => state.dokumenter.data,
```
Kan bruke `RootState` fra `AppTypes` her

sigbjorn-myhre replied:
Ser at jeg har glemt å endre denne fra før jeg pullet dine endringer 😅

JohanSakshaug commented on src/ducks/dokumenter/selectors.ts:
```diff
@@ -0,0 +1,87 @@
+
+export const DokumentOversiktSelector = createSelector(
+  dokumenterSelector,
+  (dokumenter: { dokumentOversikt: DokumentOversikt[] }) => dokumenter.dokumentOversikt || []
```
Typen til `dokumenter` skal bli inferred når `dokumenterSelector` bruker `RootState`. Dette gjelder for alle selectorene i denne fila.

JohanSakshaug commented on src/felleskomponenter/sideDialog/VedleggVelger.tsx:
```diff
@@ -0,0 +1,212 @@
+import * as Ikoner from '../../resources/images';
+
+import PdfLink, { lagPdfUrl } from './pdfLink';
+import { formatterDatoTilNorsk } from '../../utils/dato';
```
Kan bruke `* as Utils` for konvensjonens skyld

JohanSakshaug commented on src/felleskomponenter/sideDialog/VedleggVelger.tsx:
```diff
@@ -0,0 +1,212 @@
+      </td>
+      <td>
+        { redigerer
+          ? <span />
```
Kan returnere `null` i stedet for `<span />` her for tydelighet

JohanSakshaug replied:
Når jeg ser på den igjen, så tenker jeg `!redigerer && <Mui.Knapp />` er enda klarere 😄

sigbjorn-myhre replied:
Så denne nå. Enig der! 👍

JohanSakshaug commented on src/felleskomponenter/sideDialog/VedleggVelger.tsx:
```diff
@@ -0,0 +1,212 @@
+            <tr>
+              <td colSpan={6}>
+                <Nav.Checkbox
+                  onChange={e => markerAlleVedlegg(e)}
```
`onChange={markerAlleVedlegg}`

JohanSakshaug commented on src/felleskomponenter/sideDialog/VedleggVelger.tsx:
```diff
@@ -0,0 +1,212 @@
+
+  const leggTilMarkerteVedleggHandler = () => {
+    setValgteVedlegg(alleVedlegg.filter(v => markerteVedlegg.includes(v.id)));
+    toggleRedigerer();
```
En annen måte man kunne gjort denne togglen på er å ta et prop `onLeggTil` som kalles i `leggTilMarkerteVedleggHandler`. `toggleredigerer` kan da oppgis i foreldre-komponenten med dette propet. Da slipper VedleggListe å vite om togglen.

Samme ved avbrytHandler.

JohanSakshaug replied:
Dette burde vel også kunne gjøres for setValgteVedlegg såvidt jeg kan se

JohanSakshaug commented on src/felleskomponenter/ui/knapp.tsx:
```diff
@@ -1,18 +1,21 @@
 import React from 'react';
 import PT from 'prop-types';
+import { KnappBaseProps } from 'nav-frontend-knapper';
```
👍

JohanSakshaug commented on src/services/modules/dokumenter/dokument.ts:
```diff
@@ -0,0 +1,15 @@
+import { getAsJson, postAsJson } from '../../utils';
+import { API_BASE_URL, DOKUMENTER } from '../../api-constants';
+
+export type BrevbestillingDto = {
```
`BrevbestillingReqDto` som vi diskuterte. Og `DokumentOversiktResDto`

JohanSakshaug replied:
Jeg lager PR på Dto'er som ligger på develop

JohanSakshaug commented on src/services/utils.js:
```diff
@@ -29,11 +29,14 @@ const toJson = async response => {
 };
 
-export const sendResultatTilDispatch = (dispatch, action, callback) => (...data) => {
+export const sendResultatTilDispatch = (dispatch, action, callback, mapDispatchData) => (...data) => {
```
Synes det blir litt uklar navngiving for callback her. Kunne vi samlet callback og mapDispatchData under et objekt som heter callbacks(slik som i doThenDispatch)? Og endret callback => onDispatch?

`sendResultatTilDispatch = (dispatch, action, { onDispatch, mapDispatchData })`

sigbjorn-myhre replied:
Det høres ut som en god løsning 👍

sigbjorn-myhre committed:
QA MELOSYS-3535

- Legger til tester for dokumenter/selectors
- Oppdaterer typer på dokumenter/selectors

sigbjorn-myhre committed:
redigerer-ternary til !redigerer &&

sigbjorn-myhre committed:
Dato er nullable
